### DISCUSSION TITLE
Trouble With cl.CreateImage

### DISCUSSION THREAD

**Discussion by Squaduck** [dotnet/Silk.NET#1762] [Q&A] [answered]
Hello! I've been trying to add OpenCL to my application, and I can't get cl.CreateImage to work.
I have no prior experience with OpenCL and haven't found any relevant tutorials.
I've scoured the specifications and dug for examples, but I can't find anything about making a 2D Image array.
```C#
public int PrepareOutputImageArr(int width, int height, int num_imgs)
        {
            ImageFormat imageFormat = new(ChannelOrder.Rgb, ChannelType.UnsignedInt8);
            ImageDesc imageDesc = new()
            {
                ImageType       = MemObjectType.Image2DArray,
                ImageWidth      = (nuint)width,
                ImageHeight     = (nuint)height,
                ImageDepth      = 0,
                ImageArraySize  = (nuint)num_imgs,
                ImageRowPitch   = 0,
                ImageSlicePitch = 0,
                NumMipLevels    = 0,
                NumSamples      = 0,
            };
            out_img_arr = cl.CreateImage(context, MemFlags.HostReadOnly | MemFlags.WriteOnly, imageFormat, imageDesc, null, out int err);
            if (CheckErr(err, "cl.CreateImage") != 0)
                return 1;
            return 0;
        }
```
This isn't my exact code, but it's pretty much this with different variable names.
As far as I can tell, this should work. But, it always throws an InvalidImageDescriptor error code.
I've tried setting all the 0s to null, 1, or even not setting them at all, but to no avail. 
The sizes are all well within limits.
I've tried making an empty array to point to, but that didn't help either.
From what I can tell, this is exactly to spec. eg, If host_ptr is null, don't set pitch or certain mem flags, etc.

I also tried creating a 2D image to get the hang of it, but I couldn't even get that to work.

I'm using a GeForce 1660 SUPER, and the platform version returns OpenCL 3.0 CUDA 12.2.146.
My computer is running OpenSUSE Tumbleweed (x86_64) with official Nvidia drivers and Cuda libraries.
I've successfully ran a hello world program, and I'm using the latest release of Silk.NET (2.18.0)

I know that this isn't stack overflow and maybe not the right place for questions like this, but I'd rather not risk having my honest question marked as duplicate.

**Comment by qian-o** [comment]
What is the return value of err?

**Reply by qian-o** [reply]
I have also tried using this function to create Image2D, but it will output CL_INVALID_OPERATION exception.

There is no problem with 'CreateImage2D' with the same parameter configuration.

**Reply by Squaduck** [reply]
err's return value is -65 (invalid image descriptor) 

It's reassuring to hear that other people are also having trouble with this function because I thought I might have just missed something really simple.
I also haven't had any issues with the CreateImage2D function.

**Comment by Squaduck** [comment]
I wonder if this could be because the Silk.NET `ImageDesc` struct doesn't have the `mem_object`/`buffer` member? Looking in [the docs](https://registry.khronos.org/OpenCL/sdk/3.0/docs/man/html/cl_image_desc.html) (or [the specifications](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#cl_image_desc)) it seems Silk.NET is missing a `cl_mem` type member.

I have no clue how `union`s work, or if C# has anything close, but this is my only real idea.

My only full theory is that `clCreateImage` is expecting everything in the `ImageDesc` to be set to something, but the `cl_mem` type member isn't properly set to anything, and thus the image descriptor in its entirety is invalid. Not that I think this theory is of use to anyone, I just want to make my thought process clear.

**Reply by Perksey** [reply]
Great spot! That’s probably it, I’ll look into how this differs with Vulcan and see if this can get fixed :)

**Reply by qian-o** [reply]
Wow, it turns out that way. I tried OpenTK and it had no issues.

**Reply by qian-o** [reply]
I had considered this issue before, but I felt like it was being eliminated so I didn't pay much attention to it.

**Reply by Perksey** [reply]
This should hopefully be an easy fix, will probably just route the OpenCL struct handling through to the Vulkan backend in the generator which we know works with unions (unless we’re doing that already, in which case I need to get more creative haha)

**Reply by Perksey** [reply, marked as answer]
Ah so it turns out we did do that already, but it's not working because Khronos made no attempt to be consistent with Vulkan here:
```xml

            <!-- Can't properly express this in XML schema yet - use name/ tag to align? -->
            <member>union {
        cl_mem buffer;
        cl_mem mem_object;
    }</member>
```
Will manually hack this in.

**Comment by Squaduck** [comment]
I see that the latest release addresses this. I will try it when I have time.
